### ISSUE TITLE
Assistance1

### ISSUE THREAD

**Issue by harmongt01** [simat/BatteryMonitor#2]
Hi,

The project is very close to what I'm trying to do with the same hardware.  Imported your code onto the BeagleBone Black, which went well thanks to your clear instructions and nice work.  I've reviewed your descriptions and schematic but I need a little help to modify for my project.  

Here's an example of the monitor points:
12V battery bank, voltage of entire bank.  Reading 1
12V current shunts on incoming power from solar panels.  Reading 2
12V current shunts on incoming power from wind generator1.  Reading 3
12V current shunts on incoming power from wind generator2.  Reading 4
12V current shunts on outgoing power from battery bank to inverter.  Reading 5

It looks like from the schematic you did not use a power divider for the Bat Current, is that correct?

I am new to this so I'd like to use work you've already done.  Tweak it for the above readings and go from there.

Thanks in advance,
George

**Comment by harmongt01** [comment]
Tried editing the battery.cfg to what I thought looked ok for on current monitor point but am getting this error:
root@beaglebone:~/Adafruit_Python_ADS1x15/BatteryMonitor# python
Python 2.7.9 (default, Aug 13 2016, 17:56:53) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import batteries
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "batteries.py", line 32, in <module>
    config['alarms'][i][1] = compile(config['alarms'][i][1], '<string>', 'exec') 
TypeError: expected a readable buffer object

How can I enable a debug log?

**Comment by harmongt01** [comment]
Made some progress:
>>> import batteries
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "batteries.py", line 27, in <module>
    from getdata import Readings
  File "getdata.py", line 32, in <module>
    class Readings:
  File "getdata.py", line 40, in Readings
    avoffset = avoffset/numcells
ZeroDivisionError: float division by zero
>>> batteries.deamon()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'batteries' is not defined

**Comment by simat** [comment]
Can you post your battery.cfg file

**Comment by harmongt01** [comment]
[battery.txt](https://github.com/simat/BatteryMonitor/files/818297/battery.txt)
 
I changed the ext to .txt so the site would accept it.

**Comment by simat** [comment]
numcells should be 1.

**Comment by harmongt01** [comment]
ok.  I was trying to one shunt meter using posinon AIN0 and AIN1 of only one ADS1115 to start with.

**Comment by harmongt01** [comment]
My message was not clear, sorry.  What I meant to say is I was trying to start with one shunt and one ADS1115 to take a current measurement using differential inputs.

Then when that works I will add the others as well as a voltage measurement of the battery bank.

**Comment by simat** [comment]
When I wrote the code I never thought about not monitoring battery voltage. I would just set up one dummy voltage input.

The code could be altered ...

**Comment by harmongt01** [comment]
Thanks simat.
Makes sense.  The voltage is required to calculate the current.
So I'll play with it and see what I can figure out.

**Comment by simat** [comment]
As you say the current is calculated by measuring the voltage across the current shunts and converting these to a current. These voltage inputs are defined by the [Current Inputs] section of battery.cfg.

The primary reason for this project was to measure the voltages on each of the individual cells in a lithium battery. Current monitoring was a secondary objective.

SImon

**Comment by simat** [comment]
Hi George,
I have been thinking further about this. Currently my software handles a variable number of voltage inputs and two current inputs.

What you want is four current inputs. It wouldn't be too difficult to change the software to take a variable number of current inputs. The number of inputs would be defined in the [Current Inputs] section of the config file.

I am not sure what programming skills you have. Do you want to try to change the code with help from me or would you like me to do it?

Simon

**Comment by harmongt01** [comment]
Hi Simon,
Can you help me to understand these current cfg lines.  chP, chN, pga and sps?
[CurrentInputs]
#ibat = 'atod2.readADCDifferential(chP=1, chN=2, pga=2048, sps=250)'
#iin = 'atod2.readADCDifferential(chP=2, chN=3, pga=2048, sps=250)'

I have added one battery voltage monitor point with voltage divider.  The voltage will be at channel=0 using SingleEnded measurement.
I would like to use the second two channels for a Differential reading from a current shunt.
Without the current positions enabled I still get the following error on 'import batteries'.  I probably have something connected wrong but wanted to be sure this cfg should give me a single voltage reading.

>>> import batteries
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "batteries.py", line 32, in <module>
    config['alarms'][i][1] = compile(config['alarms'][i][1], '<string>', 'exec') 
TypeError: expected a readable buffer object

**Comment by harmongt01** [comment]
Hi Simon,

Just saw your last post (must not have hit the refresh).  I don't have the programming experience but would like to get a base code working that I can learn from.  And would be willing to pay you for your time to assist.  You can email me at harmongt@live.com.

George

**Comment by simat** [comment]
Hi George,

I am happy to help, if we continue to do it through Github it may help others as well.

> Can you help me to understand these current cfg lines. chP, chN, pga and sps?
[CurrentInputs]
#ibat = 'atod2.readADCDifferential(chP=1, chN=2, pga=2048, sps=250)'
#iin = 'atod2.readADCDifferential(chP=2, chN=3, pga=2048, sps=250)'

In this statement
>   self.rawcurrent = eval(config['CurrentInputs']['ibat']) # Battery current in counts

in the module getdata.py my program executes the code you specify in the [CurrentInputs] section of the config file to get the differential voltage reading which equates to the current.

In this case  'atod2.readADCDifferential()' is a function in the Adafruit driver ADS1x15.py. The arguments passed to the function are contained in the brackets. chP=x is the channel number of the positive differential input, chN=y is the negative channel, pga=a sets the amplifier gain of the A-D converter and sps=b is the number of samples per second.

For two current inputs from two 50mV shunts the best setup would be

ibat = 'atod2.readADCDifferential(chP=1, chN=3, pga=256, sps=250)'
iin = 'atod2.readADCDifferential(chP=2, chN=3, pga=256, sps=250)'

pga=256 gives a voltage span of 256mV. It is best to use channel 3 as the common negative channel as it means you can have three differential inputs.

For your voltage input
v1 = 'atod0.readADCSingleEnded(channel=0, pga=2048, sps=250)'

Here is a link to the data manual for the ADS1115, [https://www.google.com.au/url?sa=t&rct=j&q=&esrc=s&source=web&cd=2&ved=0ahUKEwjr3u3P3bvSAhWFFpQKHecRCPgQFgggMAE&url=http%3A%2F%2Fwww.ti.com%2Flit%2Fds%2Fsymlink%2Fads1115.pdf&usg=AFQjCNHQ7BsG5xWhHSCJDDl9uvuSIcA3mQ&sig2=m1bk-fMmHNngutbux6vPcQ](url)

Simon

**Comment by harmongt01** [comment]
Hi Simon,

Thank you for the explanation.

updated my config but still see this error:
>>> import batteries
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "batteries.py", line 32, in <module>
    config['alarms'][i][1] = compile(config['alarms'][i][1], '<string>', 'exec') 
TypeError: expected a readable buffer object

George

**Comment by harmongt01** [comment]
The link didn't work for me but I did find ADS1115.pdf from Texas instuments.

**Comment by simat** [comment]
Hi George,

Sorry my mistake, I have change the code that checks the alarms but haven't updated the default config files. Replace the line in the [alarms] section of the config file with this.

alarm = ['GPIO.setup("P8_10", GPIO.OUT)', 'test=minvolts<2.8 or maxvolts>3.60', 'GPIO.output("P8_10", GPIO.HIGH)', \\
        'test=minvolts>2.9 and maxvolts<3.59', 'GPIO.output("P8_10", GPIO.LOW)']

**Comment by simat** [comment]
Other option is to just comment out the alarm.

I have updated the default config files

**Comment by simat** [comment]
I have fixed the link to the ADS1115 documentation

**Comment by harmongt01** [comment]
Hi Simon,

after replacing the line in the alarms section I get this error:
>>> import batteries
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "batteries.py", line 24, in <module>
    from config import config
  File "config.py", line 31, in <module>
    config = Config()
  File "config.py", line 28, in __init__
    self.config[section][key] = literal_eval(val)
  File "/usr/lib/python2.7/ast.py", line 49, in literal_eval
    node_or_string = parse(node_or_string, mode='eval')
  File "/usr/lib/python2.7/ast.py", line 37, in parse
    return compile(source, filename, mode, PyCF_ONLY_AST)
  File "<unknown>", line 1
    ['GPIO.setup("P8_10",GPIO.OUT)','test=minvolts<2.8 or maxvolts>3.60','GPIO.output("P8_10",GPIO.HIGH)',\'test=minvolts>2.9 and maxvolts<3.59','GPIO.output("P8_10",GPIO.LOW)']
                                                                                                                                                                                ^
SyntaxError: unexpected character after line continuation character

------------------------------
Then when I comment it out, the import batteries works but when I try to run the deamon it throws this error:
Python 2.7.9 (default, Aug 13 2016, 17:56:53) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import batteries
>>> batteries.deamon()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "batteries.py", line 41, in deamon
    logsummary = summary.Summary()
  File "summary.py", line 61, in __init__
    if self.summary['yeartodate']['timestamp'][0:4] != printtime[0:4]:
KeyError: 'timestamp'

**Comment by simat** [comment]
On the line after the [yeartodate] label in the summary file add this line
timestamp = '20140101000000'

**Comment by simat** [comment]
> ['GPIO.setup("P8_10",GPIO.OUT)','test=minvolts<2.8 or maxvolts>3.60','GPIO.output("P8_10",GPIO.HIGH)','test=minvolts>2.9 and maxvolts<3.59','GPIO.output("P8_10",GPIO.LOW)']

I don't think the two lines copied and pasted properly.

If you do git pull you will get the latest updated sample config files and the fixed summary file.

**Comment by harmongt01** [comment]
how should I do the git pull?
from the BatteryMonitor directory run this command?  
git clone https://github.com/simat/BatteryMonitor.git

**Comment by harmongt01** [comment]
ok.. just git pull
I'm new at this.

**Comment by harmongt01** [comment]
Ok.. still moving forward 👍 

getting this error:
>>> import batteries
>>> batteries.deamon()
Error accessing 0x48: Check your I2C address
Error accessing 0x48: Check your I2C address
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "batteries.py", line 49, in deamon
    batdata = Readings()  # initialise batdata after we have valid sys time
  File "getdata.py", line 104, in __init__
    self.getatod()
  File "getdata.py", line 86, in getatod
    self.rawvolts[i+1] = eval(vin[i])/1000 # A to D 1 to 4 in volts
  File "<string>", line 1, in <module>
  File "ADS1x15.py", line 223, in readADCSingleEnded
    val = (result[0] << 8) | (result[1])
TypeError: 'int' object has no attribute '__getitem__'

**Comment by harmongt01** [comment]
if no pinouts are connected to assign the ADS address and I have only one ADS1115 then the default should be the 0x48 right?

**Comment by simat** [comment]
Do you have pullup resistors on the I2C clock (scl) and data (sda) lines. you need a 10kOhm resistor between these lines and Vdd (3.3V)

**Comment by simat** [comment]
Pin 1 (addr) need to be connected to GND

**Comment by harmongt01** [comment]
Between the ADS (scl/sda) to BBB (scl/sda) I do not have 10kOhm resistors.. 
Between the ADS VDD and the BBB ADS_VDD I do not have 10kOhm resistor either.  I do not see that in the schematic.  I will add those.  
P9 pin1 BBB should be connected to GND of the battery bank?

Hope I didn't break it....

**Comment by simat** [comment]
The pullup resistors need to be between scl and Vdd and sda and Vdd, not between the ADS and BBB.

If your BBB is being powered by a non isolated power supply that is being powered from your battery the GND will be connected via the power supply.

**Comment by harmongt01** [comment]
Ok got it.  Will make the changes.  
The BBB is powered by an AC power supply adaptor for now.

**Comment by harmongt01** [comment]
Hi Simon,

I will try this tomorrow.  Thank you very much for your help!

-George

**Comment by simat** [comment]
The AC power supply will be isolated from the battery so you will need to connect the battery - to P9 pin 1.
I checked my circuit diagram and have realised that the BatteryMonitor.pdf circuit diagram is an old out of date one. I have uploaded a new one.

As you are not monitoring all the individual cell voltages your circuit will be quite simple. Can you upload your circuit so I can see what you have come up with.

Simon

**Comment by harmongt01** [comment]
![20170307_161018](https://cloud.githubusercontent.com/assets/26151742/23680850/561cc4da-0352-11e7-8969-0965f3f60e78.jpg)
Here's a photo but it's really hard to see what's connected to where.
I still have the same issue so I will pull out the connections and start over.  Do you think I should see some kind of output if I have no voltage or shunts connected?

**Comment by simat** [comment]
You shouldn't be getting any errors and be getting some kind of output without any inputs connected.

I see you have two ADS1115s connected. Is each one on a separate address? The one with address 0x48 should have pin 1 (ADDR) connected to GND and the second one (address 0x49) should be at Vdd (3.3V)

To get some sort of output form the ADS1115s all you need to connect is 
Pin 1 = GND
Pin 3 = GND
Pin 8 = Vdd (3.3V)
Pin 9 = SDA Beagle P9-20
Pin 10 = SCL Beagle P9-19

Check these connections first and if you cant get anywhere I suggest you strip your circuit down to the bare minimum and see if it works.

Simon

**Comment by harmongt01** [comment]
Yes I have two ADS1115.  Right now only the one on the right is connected to the BBB.  The jumper from ADDR is connected to GND.

I will have to strip it down to the minimum as you suggested.

**Comment by simat** [comment]
There is a linux package i2c-tools that has a program called i2c-detect
If you run i2c-detect -r 1 you should get the following with one ADS1115

> WARNING! This program can confuse your I2C bus, cause data loss and worse!
I will probe file /dev/i2c-1 using read byte commands.
I will probe address range 0x03-0x77.
Continue? [Y/n] y
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:          -- -- -- -- -- -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
40: -- -- -- -- -- -- -- -- 48 -- -- -- -- -- -- -- 
50: -- -- -- -- UU UU UU UU -- -- -- -- -- -- -- -- 
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
70: -- -- -- -- -- -- -- --

**Comment by harmongt01** [comment]
I have it stripped down.  
Trying to run the detect program:

root@beaglebone:~# apt-get install i2c-tools
Reading package lists... Done
Building dependency tree       
Reading state information... Done
i2c-tools is already the newest version.
0 upgraded, 0 newly installed, 0 to remove and 21 not upgraded.
root@beaglebone:~# i2c-detect -r 1
bash: i2c-detect: command not found
root@beaglebone:~#

**Comment by harmongt01** [comment]
not sure why it has the strike through on those lines.  Not copying over correctly I guess.

**Comment by harmongt01** [comment]
I tried this:
root@beaglebone:~# i2cdetect -y 1
Error: Can't use SMBus Quick Write command on this bus

**Comment by simat** [comment]
I don't understand why i2cdetect -r 1 doesn't work
Here is the response I get to trying different options with i2cdetect

> simat@beaglebone:~$ i2cdetect
Error: No i2c-bus specified!
Usage: i2cdetect [-y] [-a] [-q|-r] I2CBUS [FIRST LAST]
       i2cdetect -F I2CBUS
       i2cdetect -l
  I2CBUS is an integer or an I2C bus name
  If provided, FIRST and LAST limit the probing range.

> simat@beaglebone:~$ i2cdetect -l
i2c-0	i2c       	OMAP I2C adapter                	I2C adapter
i2c-1	i2c       	OMAP I2C adapter                	I2C adapter

> simat@beaglebone:~/BatteryMonitor$ i2cdetect -r 1
WARNING! This program can confuse your I2C bus, cause data loss and worse!
I will probe file /dev/i2c-1 using read byte commands.
I will probe address range 0x03-0x77.
Continue? [Y/n] y
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:          -- -- -- -- -- -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
40: -- -- -- -- -- -- -- -- 48 49 4a -- -- -- -- -- 
50: -- -- -- -- UU UU UU UU -- -- -- -- -- -- -- -- 
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
70: -- -- -- -- -- -- -- --    

> simat@beaglebone:~$ i2cdetect -y 1
Error: Can't use SMBus Quick Write command on this bus

I would try the i2cdetect -l first to see if the i2c interfaces are working then the i2cdetect -r 1
                     
Hope this is useful, Simon

**Comment by simat** [comment]
Hi George,
Had a bit of spare time and looked at your circuit in more detail.
The Adafruit boards have the 10k pullups for SCL and SDA and a pull down for ADDR so you could leave ADDR disconnected and have no external pullups for the circuit to work at address 0x48.

Would you like me to look at the circuit diagram of your current sensing and battery voltage divider circuits?

SImon

**Comment by harmongt01** [comment]
HI Simon,

I ran those commands:
root@beaglebone:~# i2cdetect -l
i2c-0   i2c             OMAP I2C adapter                        I2C adapter
i2c-1   i2c             OMAP I2C adapter                        I2C adapter
i2c-2   i2c             OMAP I2C adapter                        I2C adapter

root@beaglebone:~# i2cdetect -r 1
WARNING! This program can confuse your I2C bus, cause data loss and worse!
I will probe file /dev/i2c-1 using read byte commands.
I will probe address range 0x03-0x77.
Continue? [Y/n] Y
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:          -- -- -- -- -- -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
70: -- -- -- -- -- -- -- --                     

I have the ADS down to only the basic connections.  VDD, GND, SCL and SDA.  I removed the jumper between GND and ADDR so the address should still be 0x48.

**Comment by harmongt01** [comment]
I think my problem is down to the BBB not seeing the ADS boards.  I tried both boards one at a time and same thing.  Nothing found with i2cdetect.  

Possible problems as I see it:
1.  Either both ADS boards are bad 
2.  something is wrong with the SCL and SDA ports at the BBB
3.  SCL and SDA ports not enabled.  I read these should be enabled by default.
4.  driver for ADS boards not installed?

**Comment by harmongt01** [comment]
Hi Simon,

I don’t have any diagrams for my current sensing or voltage divider.  The voltage at the divider output point measured 2.6VDC when I first put it in.

I have two 100A 75mV shunts in place.  These are in series with the negative sides of my input from solar panels and input from small wind generators.  Wind and solar on different shunts.

I have a feeling something is not right with my BBB SCL and SDA but I don’t know how to check if the ports are ok.

Thanks,
George

**Comment by simat** [comment]
> root@beaglebone:~# i2cdetect -l
i2c-0 i2c OMAP I2C adapter I2C adapter
i2c-1 i2c OMAP I2C adapter I2C adapter
i2c-2 i2c OMAP I2C adapter I2C adapter

I see you have the new BBB with WiFi. Interestingly it has three i2c devices listed where my old BBB has two.

Try i2cdetect -r 2 and then i2cdetect -r 0

**Comment by harmongt01** [comment]
I have different results with those commands, looks more promising:

root@beaglebone:~# itcdetect -r 0
bash: itcdetect: command not found
root@beaglebone:~# i2cdetect -r 2
WARNING! This program can confuse your I2C bus, cause data loss and worse!
I will probe file /dev/i2c-2 using read byte commands.
I will probe address range 0x03-0x77.
Continue? [Y/n] Y
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:          -- -- -- -- -- -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
40: -- -- -- -- -- -- -- -- 48 -- -- -- -- -- -- -- 
50: -- -- -- -- UU UU UU UU -- -- -- -- -- -- -- -- 
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
70: -- -- -- -- -- -- -- --                         
root@beaglebone:~# i2cdetect -r 0
WARNING! This program can confuse your I2C bus, cause data loss and worse!
I will probe file /dev/i2c-0 using read byte commands.
I will probe address range 0x03-0x77.
Continue? [Y/n] Y
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:          -- -- -- -- -- -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: -- -- -- -- UU -- -- -- -- -- -- -- -- -- -- -- 
30: -- -- -- -- UU -- -- -- -- -- -- -- -- -- -- -- 
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
50: UU -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
70: UU -- -- -- -- -- -- --

**Comment by harmongt01** [comment]
How can you tell from that output there is an onboard Wifi?  I wouldn't have thought that would be a factor but makes sense.  You are good at this!

**Comment by simat** [comment]
Great!!! OK this means that your A-D is on bus 2 rather than 1.

If you change the line atod0 in the [AtoDs] section of battery.cfg to 

atod0 = 'AtoD.ADS1x15(busnum=2,address=0x48,ic=0x01, debug=True)'

my software should hopefully work!

**Comment by harmongt01** [comment]
Thanks Simon.  I updated the config.  Need to rewire the connections.. will send status shortly.

**Comment by simat** [comment]
> How can you tell from that output there is an onboard Wifi? I wouldn't have thought that would be a factor but makes sense. You are good at this!

I could see from your photo.

My earlier version of the board has an error where i2c bus 1 is connected to pins 19 and 20 even though the pins are called I2C2-SCL and I2C2-SDA. I now remember this caused me some grief when I first designed this. Looks like the fault has been fixed with your version of the BBB.

**Comment by harmongt01** [comment]
Ah yes. the picture shows the wifi wires.

I just connected things back up and when trying to start the deamon it threw this error:

root@beaglebone:~/Adafruit_Python_ADS1x15/BatteryMonitor# python
Python 2.7.9 (default, Aug 13 2016, 17:56:53) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import batteries
>>> batteries.deamon()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "batteries.py", line 49, in deamon
    batdata = Readings()  # initialise batdata after we have valid sys time
  File "getdata.py", line 104, in __init__
    self.getatod()
  File "getdata.py", line 87, in getatod
    self.rawcurrent = eval(config['CurrentInputs']['ibat']) # Battery current in counts
  File "<string>", line 1, in <module>
NameError: name 'atod2' is not defined
![20170308_212226](https://cloud.githubusercontent.com/assets/26151742/23734690/6efe5a78-0446-11e7-8b15-b802378123c9.jpg)

**Comment by harmongt01** [comment]
the voltage out going to A0 was 2.58VDC when I connected it.

**Comment by simat** [comment]
ops my mistake, can you change atod2 in the lines in the [CurrentInputs] section to atod0, this corresponds to the AtoD address you defined in the [AtoDs] section

If your voltage input is 2.58V you will need to change pga=2048 to pga=4096 in the [VoltageInput] section

**Comment by harmongt01** [comment]
Next error:

root@beaglebone:~/Adafruit_Python_ADS1x15/BatteryMonitor# python
Python 2.7.9 (default, Aug 13 2016, 17:56:53) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import batteries
>>> batteries.deamon()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "batteries.py", line 49, in deamon
    batdata = Readings()  # initialise batdata after we have valid sys time
  File "getdata.py", line 104, in __init__
    self.getatod()
  File "getdata.py", line 98, in getatod
    self.uncalvolts[i] = self.rawvolts[i]*config['calibrate']['batvgain'] # A/D to battery volts
KeyError: 'batvgain'
>>>

**Comment by harmongt01** [comment]
Just checked the reading at the voltage divider:  2.68VDC
Voltage at the battery bank is 13.23VDC.

**Comment by simat** [comment]
Add this line in [calibrate] section

batvgain = 4.94 # resistor divider gain 13.23/2.68

I will be away from the computer for a couple of hours. I am in Western Australia. The time here is 12:01pm

**Comment by harmongt01** [comment]
I'd love to explore the northwestern side of Australia some day... on my bucket list

**Comment by harmongt01** [comment]
>>> import batteries
>>> batteries.deamon()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "batteries.py", line 49, in deamon
    batdata = Readings()  # initialise batdata after we have valid sys time
  File "getdata.py", line 104, in __init__
    self.getatod()
  File "getdata.py", line 91, in getatod
    +config['calibrate']['pcurrent'] # current in amp  
KeyError: 'ibatgain'
>>>

**Comment by simat** [comment]
I think there is an error in the [calibrate] section of the config file. Can you upload it to me again
Thanks Simon

**Comment by simat** [comment]
The northwest of WA is fantastic. Large distances to travel though

**Comment by harmongt01** [comment]
It is a big country.  I like the western side of the US as well.  Lived there for almost 10yrs after my time in the service.  Now I'm back near my hometown of St. Louis.  Not as much free land to roam near big cities.  Every inch of land is owned by someone.  

I attached my config.  I may have something not set correctly.  
Thanks,

George
[battery.txt](https://github.com/simat/BatteryMonitor/files/832052/battery.txt)

**Comment by simat** [comment]
Hi George,
I think I have spotted the error,
Change ibatvgain in [calibrate] to batvgain

Simon

**Comment by harmongt01** [comment]
Hi Simon,

Little different now:
root@beaglebone:~/Adafruit_Python_ADS1x15/BatteryMonitor# python
Python 2.7.9 (default, Aug 13 2016, 17:56:53) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import batteries
>>> batteries.deamon()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "batteries.py", line 77, in deamon
    batdata.getraw()
  File "getdata.py", line 128, in getraw
    self.deltav[i]=round((self.batvoltsav[i]-self.batvoltsav[i-1]-config['calibrate']['delta'][i-1]),3)
KeyError: 'delta'

**Comment by simat** [comment]
Sorry, another change in the way I calculate things that didn't show up in the default conig files
Add 
delta = [ 0.0, 0.0, 0.0, 0.0]
to the [calibrate] section

**Comment by harmongt01** [comment]
No worries man.  I'm still working... getting late.
Thanks again for all your help!

Oh yeah!!!  We have data:
>>> import batteries
>>> batteries.deamon()
11.691 11.69 6.691 0.0 0.0 10000.0 10000.0
11.691 11.69 6.691 0.0 0.0 10000.0 10000.0
11.691 11.69 6.691 0.0 0.0 10000.0 10000.0
11.691 11.69 6.691 0.0 0.0 10000.0 10000.0
11.691 11.69 6.691 0.0 0.0 10000.0 10000.0

**Comment by simat** [comment]
Great!
those figures are
first one is cell voltage ( in your case overall battery voltage)
second one is overall battery voltage
Third is difference in voltage between the cells which you can ignore
forth one is current 'ibat' which I measure as current in/out of the battery, in your case ain1
fith is current 'iin' which I measure as current from solar panels, in your case ain2
sixth is raw battery SOC
last is adjusted SOC taking into account the efficiency of the battery

Simon

**Comment by harmongt01** [comment]
Excellent!!
Can I remove 2, 3, 6, 7?

I'm interested in battery bank overall voltage.  All are in parallel, deep cycle marine batteries.  This is my first bank so I'm starting with a 12V system.  Wind turbines come in as AC, then rectified to 12VDC close to the battery bank.  Solar panels are coming in DC so I have fat power cables with circuit breakers coming down from those.  I will eventually add a second and third shunt to measure outgoing current to the inverters.

My first goal with this is to get historical data that I can put into graph form to see what is actually going on.  I may need to monitor the status of each battery as well but I'm not sure how to do that.  It would be good to know since one bad battery can take down the whole thing.  If they are all connected in parallel then is that possible?

**Comment by simat** [comment]
The data dumped to the terminal is really just debug information. I would use the data from the log files and the summary file for historical data gathering. The summary file is updated once a minute.

My web page generator takes information from the summary file to generate the battery status web page. I have a web page server running on the BBB.

If you are wiring batteries up in parallel it is most important to wire them up properly so that each battery shares the current equally. There is a good description of how to do this here http://www.smartgauge.co.uk/batt_con.html

If wired up correctly in parallel strong batteries can support the weak ones. If they are wired up incorrectly the load on the batteries will not all be the same so the battery with the highest load will not last as long.

The SOC figures can be useful as a guide to how much power is in your battery. They are reset back to 100% when the battery has reached the full voltage and the charge current has tapered off at the end of the absorb phase of the charging.

Are you using some sort of charge controller to control the current going into your battery from the solar panels and the turbines?

Simon

**Comment by simat** [comment]
I have added a new circuit diagram called BMSsingle.pdf which should be useful for your setup. The resistors and capacitors on the inputs of the ADS1115 are important as they act as a filter.

**Comment by harmongt01** [comment]
Thanks!
Also, how can I change the summary file to be written to /media/SD/summaryfile
When I change it in the config it throws an error.

**Comment by harmongt01** [comment]
Yes I have a charge controller by Coleman Air.

**Comment by simat** [comment]
Hi George,
you will have to copy the default summary file to /media/SD/summaryfile before you run the software or you could leave it in the BatteryMonitor directory and access it there from there.
Simon

**Comment by harmongt01** [comment]
Where do I change in the config so it knows where I moved the summary file?
summaryfile = '/media/root/SD/summaryfile/'

when I added this line I get an error when I run the program.

>>> batteries.deamon()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "batteries.py", line 41, in deamon
    logsummary = summary.Summary()
  File "summary.py", line 55, in __init__
    if self.summary['hour']['timestamp'][0:10] != printtime[0:10]:
KeyError: 'hour'

**Comment by harmongt01** [comment]
I see it is creating new summary files in the BatteryMonitor directory.  I would like to write all data and logging onto the SD card since it has a lot more available space.

**Comment by simat** [comment]
Your will have to copy the default summary file to /media/SD/summaryfile before you start the porgram

**Comment by harmongt01** [comment]
so I moved it from the root of BatteryMonitor now to /media/SD/summaryfile.  When I start the problem I get the previous error
KeyError: 'hour'

I must have done something wrong.  I know it's early where you are so no rush right now.  I'm at work so can connect and look anytime I have a moment.

Thanks,
George

**Comment by simat** [comment]
I have created a new summary file called summary.1cell. If you do a git pull it should show up in your directory. Copy that across to summaryfile on your SD card.

Hopefully I haven't made any mistakes. I can't test anything as I don't have a test system, only the two systems that are in use. Maybe I should make one up ...

**Comment by harmongt01** [comment]
It did't pull that file for some reason.

**Comment by harmongt01** [comment]
HI Simon,

How will the program know to read summary from the /media/SD/summaryfile?  The location I defined in the cfg?

**Comment by harmongt01** [comment]
[battery.txt](https://github.com/simat/BatteryMonitor/files/835263/battery.txt)


I must have fubarred something because I just keep getting this error:
batteries.deamon()
Traceback (most recent call last):
File "", line 1, in 
File "batteries.py", line 41, in deamon
logsummary = summary.Summary()
File "summary.py", line 55, in init
if self.summary['hour']['timestamp'][0:10] != printtime[0:10]:
KeyError: 'hour'

**Comment by harmongt01** [comment]
Hi Simon,

Got it...(sort of)
in the [files] section of the config
summaryfile = '/media/root/SD/summaryfile'
will not work.  It is only working when I have the summary file in the BatteryMonitor directory I am running from and summaryfile = 'summary'

I must be misunderstanding how to move where it writes the summaryfile.  I'll be away for my daughters 21st birthday this evening.  I will probably continue looking into this tomorrow sometime.

Thanks,
George

**Comment by simat** [comment]
> It did't pull that file for some reason.

did you get an output something like this

> simat@simatdebian:~/beaglebone/BatteryMonitor$ git pull
remote: Counting objects: 35, done.
remote: Compressing objects: 100% (26/26), done.
remote: Total 35 (delta 17), reused 6 (delta 6), pack-reused 3
Unpacking objects: 100% (35/35), done.
From https://github.com/simat/BatteryMonitor
   f8ae8a7..9a25a4c  master     -> origin/master
Updating f8ae8a7..9a25a4c
Fast-forward
 BMSsingle.pdf                  | Bin 0 -> 44365 bytes
 BatteryMonitor.pdf             | Bin 117278 -> 167686 bytes
 battery16cell.cfg              |  11 ++++----
 battery4cell.cfg               |   9 ++++---
 battery8cell.cfg               |  16 ++++--------
 summary.4cell => summary.1cell |  57 +++++++++++++++++++++--------------------
 6 files changed, 46 insertions(+), 47 deletions(-)
 create mode 100644 BMSsingle.pdf
 rename summary.4cell => summary.1cell (55%)

Try doing another git pull

**Comment by simat** [comment]
if summary.1cell shows up in your BatteryMonitor directory try copying it with

cp summary.1cell /media/SD/summaryfile

check that the files are the same with

diff summary.1cell /medial/SD/summaryfile

If the files are the same you wont get any output from diff
Your config file should have 

summaryfile = '/media/root/SD/summaryfile'

in the [files] section.

In the last battery config file you sent to me there is a mistake in this line
logfile = '/media/root/SD//logfile/logfile.txt'
should be
logfile = '/media/root/SD/logfile/logfile.txt'

Simon

**Comment by harmongt01** [comment]
ok Simon.

That works!

I corrected the mistake in the line also.

Will take some time to monitor the readings and see what I get.

Thanks,
George

**Comment by harmongt01** [comment]
Hi Simon,

Looking at your schematic BMSsingle.pdf

What resistance is needed at R1 and R2?

**Comment by harmongt01** [comment]
Your schematics are very nice btw.  I'm not an electrical engineer so I don't have access to programs like Visio or whatever you may be using.  Is there a free program I can use?

**Comment by simat** [comment]
Hi George,
Great to see you have got it working, well done. It is always the hardest bit to get all the low level stuff working.

R1 and R2 make up the resistor divider. Vout=Vin*(R1+R2)/R2

I have a very ancient version of Eagle PCB layout program. You can get a cut down free version of the program. Looks like it has been taken over by Autocad. Here is the webpage for it http://www.autodesk.com/products/eagle/overview

I will close this issue thread so it doesn't get too long and unreadable. Feel free to open a new issue thread with any other questions. Best to try and keep one topic per thread.

Simon
